File: Sources/CountriesFlagsObjC/source/Countries/UIImage+China.m
//
//  UIImage+China.m
//  CountriesFlags
//

#import "Countries/UIImage+China.h"

#import "Extensions/UIBezierPath+Star.h"
#import "Extensions/UIBezierPath+AffineTransform.h"

@implementation UIImage (China)

+ (UIImage * _Nullable)flagChinaWithSize:(CGSize)size {
    UIColor* colorRed = [UIColor colorWithRed:238.0f/255.0f
                                        green:28.0f/255.0f
                                         blue:37.0f/255.0f
                                        alpha:1.0f];
    UIColor* colorYellow = [UIColor yellowColor];
    
    UIGraphicsBeginImageContextWithOptions(size, YES, 0.0f);
    
    [colorRed setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorYellow setFill];
    [[UIBezierPath starAtPoint:CGPointMake(size.height / 4.0f, size.height / 4.0f)
                        radius:size.height / 20.0f * 6.0f / 2.0f
                          type:FlagStarTypeFivePointed] fill];

    UIBezierPath* path = [UIBezierPath starAtPoint:CGPointMake(size.height / 20.0f * 12.2f, size.height / 20.0f * 5.45f)
                                            radius:size.height / 20.0f
                                              type:FlagStarTypeFivePointed];
    [path rotateToAngle:20.0f center:CGPointMake(size.height / 20.0f, size.height / 20.0f)];
    [path fill];

    [[UIBezierPath starAtPoint:CGPointMake(size.height / 20.0f * 12.0f, size.height / 20.0f * 7.0f)
                        radius:size.height / 20.0f
                          type:FlagStarTypeFivePointed] fill];

    path = [UIBezierPath starAtPoint:CGPointMake(size.height / 20.0f * 10.1f, size.height / 20.0f * 11.4f)
                              radius:size.height / 20.0f
                                type:FlagStarTypeFivePointed];
    [path rotateToAngle:45.0f center:CGPointMake(size.height, size.height / 20.0f * 10.0f)];
    [path fill];

    path = [UIBezierPath starAtPoint:CGPointMake(size.height / 20.0f * 10.05f, size.height / 20.0f * 6.25f)
                              radius:size.height / 20.0f
                                type:FlagStarTypeFivePointed];
    [path rotateToAngle:24.0 center:CGPointMake(size.height, size.height / 20.0f * 4.0f)];
    [path fill];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image;
}

@end

File: Sources/CountriesFlags/Countries/UIImage+China.swift
//
//  UIImage+China.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_China "wikipedia"
     
     Generates the flag of China.
     For more information see [wikipedia].
     
     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of China.
     */
    internal static func flagChina(with size: CGSize) -> UIImage? {
        let colorRed = UIColor(red: 238.0 / 255.0,
                               green: 28.0 / 255.0,
                               blue: 37.0 / 255.0,
                               alpha: 1.0)
        let colorYellow = UIColor.yellow

        UIGraphicsBeginImageContextWithOptions(size, false, 0.0)

        colorRed.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorYellow.setFill()
        UIBezierPath.star(with: CGSize(width: size.height / 2.0, height: size.height / 2.0),
                          radius: size.height / 20.0 * 6.0 / 2.0).fill()

        var path = UIBezierPath.star(with: CGSize(width: size.height / 20.0 * 24.4, height: size.height / 20.0 * 10.9),
                                     radius: size.height / 20.0)
        path.rotate(to: 20.0, around: CGPoint(x: size.height / 20.0, y: size.height / 20.0))
        path.fill()

        UIBezierPath.star(with: CGSize(width: size.height / 20.0 * 24.0, height: size.height / 20.0 * 14.0),
                          radius: size.height / 20.0).fill()

        path = UIBezierPath.star(with: CGSize(width: size.height / 20.0 * 20.2, height: size.height / 20.0 * 22.8),
                                 radius: size.height / 20.0)
        path.rotate(to: 45.0, around: CGPoint(x: size.height, y: size.height / 20.0 * 10.0))
        path.fill()

        path = UIBezierPath.star(with: CGSize(width: size.height / 20.0 * 20.1, height: size.height / 20.0 * 12.5),
                                 radius: size.height / 20.0)
        path.rotate(to: 24.0, around: CGPoint(x: size.height, y: size.height / 20.0 * 4.0))
        path.fill()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image
    }
}

#endif
